x-pack/filebeat/input/s3/input.go

```diff
@@ -178,30 +178,45 @@ func (p *s3Input) Run() {
 		if err != nil {
 			p.logger.Errorf("failed to get region name from queueURL: %v", p.config.QueueURL)
 		}
-
-		awsConfig := p.awsConfig.Copy()
-		awsConfig.Region = regionName
-
-		svcSQS := sqs.New(awscommon.EnrichAWSConfigWithEndpoint(p.config.AwsConfig.Endpoint, "sqs", regionName, awsConfig))
-		svcS3 := s3.New(awscommon.EnrichAWSConfigWithEndpoint(p.config.AwsConfig.Endpoint, "s3", regionName, awsConfig))
+		p.awsConfig.Region = regionName
 
 		p.workerWg.Add(1)
-		go p.run(svcSQS, svcS3, visibilityTimeout)
+		go p.run(visibilityTimeout)
 		p.workerWg.Done()
 	})
 }
 
-func (p *s3Input) run(svcSQS sqsiface.ClientAPI, svcS3 s3iface.ClientAPI, visibilityTimeout int64) {
+func (p *s3Input) run(visibilityTimeout int64) {
 	defer p.logger.Infof("s3 input worker for '%v' has stopped.", p.config.QueueURL)
 
 	p.logger.Infof("s3 input worker has started. with queueURL: %v", p.config.QueueURL)
+	svcSQS := sqs.New(awscommon.EnrichAWSConfigWithEndpoint(p.config.AwsConfig.Endpoint, "sqs", p.awsConfig.Region, p.awsConfig))
+
 	for p.context.Err() == nil {
 		// receive messages from sqs
 		output, err := p.receiveMessage(svcSQS, visibilityTimeout)
 		if err != nil {
-			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == awssdk.ErrCodeRequestCanceled {
-				continue
+			if awsErr, ok := err.(awserr.Error); ok {
+				if awsErr.Code() == awssdk.ErrCodeRequestCanceled {
+					continue
+				}
+
+				if awsErr.Code() == "InvalidClientTokenId" || awsErr.Code() == "ExpiredToken" {
+					p.logger.Warn(errors.Wrap(err, "credentials are invalid, acquiring AWS credentials again from config"))
+
+					awsConfig, err := awscommon.GetAWSCredentials(p.config.AwsConfig)
```

Comment:
is it possible to predict when the credentials will expire instead of waiting for the error?

Reply:
I haven't found a way to proactively check this. With the shared credential profile, we lose the expiration timestamp information. The expiration timestamp is generated when we call the `GetSessionTokenRequest` manually to generate a temporary credential. But this info is not stored in the shared credential profile. So after this `awscommon.GetAWSCredentials` function, the awsconfig shows `Expires =  0001-01-01 00:00:00 +0000 UTC` and `CanExpire =  false`.